contracts/src/finance/vesting_wallet.rs

```diff
@@ -0,0 +1,621 @@
+
+    /// Calculates the amount of tokens that has already vested. Default
+    /// implementation is a linear vesting curve.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `token` - Address of the token being released.
+    /// * `timestamp` - Point in time for which to check the vested amount.
+    ///
+    /// # Errors
+    ///
+    /// If the `token` address is not a contract, then the error
+    /// [`Error::InvalidToken`] is returned.
+    ///
+    /// # Panics
+    ///
+    /// If `total_allocation * (timestamp - self.start())` exceeds `U256::MAX`.
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_erc20(
+        &mut self,
+        token: Address,
+        timestamp: u64,
+    ) -> Result<U256, Self::Error>;
+}
+
+#[public]
+impl IVestingWallet for VestingWallet {
+    type Error = Error;
+
+    #[payable]
+    fn receive_ether(&self) {}
+
+    fn owner(&self) -> Address {
+        self.ownable.owner()
+    }
+
+    fn transfer_ownership(
+        &mut self,
+        new_owner: Address,
+    ) -> Result<(), Self::Error> {
+        Ok(self.ownable.transfer_ownership(new_owner)?)
+    }
+
+    fn renounce_ownership(&mut self) -> Result<(), Self::Error> {
+        Ok(self.ownable.renounce_ownership()?)
+    }
+
+    fn start(&self) -> U256 {
+        U256::from(self._start.get())
+    }
+
+    fn duration(&self) -> U256 {
+        U256::from(self._duration.get())
+    }
+
+    fn end(&self) -> U256 {
+        // SAFETY: both `start` and `duration` are stored as u64, so they cannot
+        // exceed `U256::MAX`
+        self.start() + self.duration()
+    }
+
+    #[selector(name = "released")]
+    fn released_eth(&self) -> U256 {
+        self._released.get()
+    }
+
+    #[selector(name = "released")]
+    fn released_erc20(&self, token: Address) -> U256 {
+        self._erc20_released.get(token)
+    }
+
+    #[selector(name = "releasable")]
+    fn releasable_eth(&self) -> U256 {
+        // SAFETY: total vested amount is by definition greater than or equal to
+        // the released amount.
+        self.vested_amount_eth(block::timestamp()) - self.released_eth()
+    }
+
+    #[selector(name = "releasable")]
+    fn releasable_erc20(
+        &mut self,
+        token: Address,
+    ) -> Result<U256, Self::Error> {
+        let vested = self.vested_amount_erc20(token, block::timestamp())?;
+        // SAFETY: total vested amount is by definition greater than or equal to
+        // the released amount.
+        Ok(vested - self.released_erc20(token))
+    }
+
+    #[selector(name = "release")]
+    fn release_eth(&mut self) -> Result<(), Self::Error> {
+        let amount = self.releasable_eth();
+        let released = self
+            .released_eth()
+            .checked_add(amount)
+            .expect("should not exceed `U256::MAX` for `_released`");
+        self._released.set(released);
+
+        evm::log(EtherReleased { amount });
```

Comment:
Originally that's how I wrote this, but opted for emitting before the transfer to avoid having to manually return `Ok(())`.

Current (the result is whatever `transfer_eth` returns):

```rust
        evm::log(EtherReleased { amount });

        transfer_eth(self.ownable.owner(), amount)
            .map_err(|_| Error::ReleaseEtherFailed(ReleaseEtherFailed {}))
```

After reorder:

```rust
        transfer_eth(self.ownable.owner(), amount)
            .map_err(|_| Error::ReleaseEtherFailed(ReleaseEtherFailed {}))?;

        evm::log(EtherReleased { amount });

        Ok(())
```

I don't mind having the latter, will change